### ISSUE TITLE
Rework ResourceDir and WorkingDirectory

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#3099]
I plan some rework to adress #3095, #2976 and #2898. Since we now have several inconsistencies, I'm planning to do this sooner rather then later.

My ideas are this:

**a)** The working directory is a logical directory, accessible from `Application.Current.WorkingDirectory`, just as it is today. When an application is started from the command-line, the *directory where star is run from* will be set to the working directory. Until proven needed, I dont plan to allow it to be customized from here.

Hence, having `C:\Users\Per\Apps\app.cs`, running that from `C:\>star \users\per\apps\app.cs` will resolve the working directory using the *current directory*.

**b)** I'm always adding the working directory to the set of directories that can include static content / resources.

**c)** Anything given as a `--resourceDirectory` is always resolved *relative to the application*, i.e. to the application .exe or .cs file. So running `C:\>star --resourceDirectory="..\foo" \users\per\apps\app.cs` will resove that to `\users\per\apps\foo`.

**d)** Resource directories can also be specified as fully qualified, rooted paths, in with case I'm not doing any resolving at all.

**e)** I'm allowing multiple resource directories by means of separating each directory with a ";" (semicolon). Each such item is subject to above resolving.

**f)** I noticed that in adminstrator, we have `http://localhost:8080/staticcontentdir` to see what resource directories have been given. And in the CLI, we refer to it as "resource directory". I want as **f1)** to pick and stick to a single term and **f2)** to review the name of the command-line parameter, to have a short form and a long form.

**g)** In Visual Studio, the resource directories will be derived from the `Command-Line` text box, and the `Working directory` can be used to specify the working directory. All the same resolution patterns as described for the CLI apply.

**h)** The directory the application is run from is still always accessible from `Application.Current.FilePath`, that specifies the fully qualified path to the application entrypoint file (.exe or .cs).

@warpech @miyconst 
Can you guys advice on **f)**? I guess you have a lot of knowledge in this area from other contexts. Also, consider **b)** and let me know if you think thats reasonable or not (quiet means you're OK with that).

@malx122 @careri @andwah @alemoi @chrhol 
Please give feedback if you have anything to add, or something you dont like!

**Comment by per-samuelsson** [comment]
Regarding **f)**, I've browsed around and "static resources" seem to be the dominant umbrella term, and "resource directory" is what I've seen most. Other alternatives include "content" (Nancy and ASP seem to use that lingo).

I leaning towards naming the parameter `--resources`, with `-res` as the short alternative. I suggest we change `http://localhost:8080/staticcontentdir` in a similar fashion. Please all give feedback on this if you have any. Its really not my area of expertise.

**Comment by careri** [comment]
Looks good to me.
Just one question about H, is that the directory in which the assembly resides? (Not the weaved assembly)

**Comment by per-samuelsson** [comment]
> Just one question about H, is that the directory in which the assembly resides? (Not the weaved assembly)

Yes. The loaded/weaved shadow files are not exposed anywhere in the API's and that is by design.

```
c:\Users\careri>dir
foo.cs

c:\Users\careri>star foo.cs
```

With that, `Application.Current.FilePath` should give you `c:\Users\careri\foo.cs`.

And similarily, if you're running from within VS and you compile your application to `D:\source\Scdb2Apps\HeadsAdmin\BootApp\bin\Debug`, then `Application.Current.FilePath` should give you `D:\source\Scdb2Apps\HeadsAdmin\BootApp\bin\Debug\BootApp.exe`.

**Comment by alemoi** [comment]
Regarding B Im thinking it can be a potential security problem. I wonder if its possible so that Visual Studio extension should set the working directory as resource dir, but in other cases (unless its wwwroot as now) - user will have to specify the resource dir explicitly.

**Comment by per-samuelsson** [comment]
Thanks for the feedback, @alemoi!

> I wonder if its possible so that Visual Studio extension should set the working directory as resource dir, but in other cases (unless its wwwroot as now) - user will have to specify the resource dir explicitly.

Technically, its not a problem at all. I can implement a pattern that it works one way in VS, and another way in tools like `star.exe`. But I can't understand exactly the reason for that pattern? How would security differ between the two, do you mean?

Also, if we dont add the working directory as a resource directory, the default for any application (whereever it's started from) would be not to have any resource directory at all. Thats fine by me, but is it what our users would want?

Should I implement some convention-based thing on top of this maybe? So that for example if a `wwwroot` folder is subfolder to the folder the application exe | cs is in, we add `wwwroot` implicitly? If so, how can we turn that off if someone wants to?

**Comment by per-samuelsson** [comment]
When reviewing how the code host currently bootstrap the application in terms of resource directories, I find some smallish things I find peculiar:

First, if a resource directory is passed, [this if-hive](https://github.com/Starcounter/Starcounter/blob/develop/src/Starcounter.Apps.JsonPatch/AppsBootstrapper.cs#L411) are doing `Path.GetFullPath(webResourcesDir);` How could that ever work? From my perspective, if we get a relative path when we are in the host, *qualifiying it with the working directory of the* ***code host*** will have no success at all? Or am I missing something here?

The next thing is the code host seem to do some magic in [this next if-hive](https://github.com/Starcounter/Starcounter/blob/develop/src/Starcounter.Apps.JsonPatch/AppsBootstrapper.cs#L417), looking for either a `wwwroot` next to the given resource directory, or else look for `[resourcedir]/src/[appname]/wwwroot`. Is that something that is implicit for all apps, like a convention?

If no one is giving input on these, I'll just let them be. They will be resolved by means of the working directory if nothing is done.

All other specified resource directories, I'm qualifying on the client (where qualification should be done) and I'll just add them with `AddStaticFileDirectory`, one by one.

**Comment by alemoi** [comment]
> But I can't understand exactly the reason for that pattern? How would security differ between the two, do you mean?

Just by having an architecture when some directory is always exposed by default (one can not ban exposing, its always on) is not a good security friendly thing.

**Comment by per-samuelsson** [comment]
Update:

We've had good creative discussions about how to tweak apps bootstrapping and sew it all together. 

I'm at the point where I'm getting pretty ready to push all of this. Some more testing to do, then documentation and updating the changelog.

**Comment by per-samuelsson** [comment]
> I'm at the point where I'm getting pretty ready to push all of this.

Turns out I'm not after all, since we cant figure out really how to resolve relative paths given. We've been discussing back and forth, and cant find any real strong case for any of the alternatives. Created #3113 for that.

**Comment by malx122** [comment]
As feedback I agree that adding a static file directory without active involvement is a major security flaw.

**Comment by per-samuelsson** [comment]
> As feedback I agree that adding a static file directory without active involvement is a major security flaw.

Thanks @malx122. I've changed that. This new story will be very straightforward and simple to comprehend I think.

**Comment by per-samuelsson** [comment]
Happy to say I finally sew this together and just merged and pushed it in to the dailys. I'm gonna as @andwah to do a check so I haven't accidently destroyed anything in the client layers, and especially regarding the app store. 

The aim/plan is we release this in a new stable coming up early next week. Meanwhile, I'll document all the new behaviours.

**Comment by per-samuelsson** [comment]
I've started documenting these changes over at [starcounter.io](http://starcounter.io/guides/network/static-file-server/). I'll eventually add a simple sheet cheet to this issue to illustrate whats been done for those not keen on docs.

**Comment by per-samuelsson** [comment]
>  I'll eventually add a simple sheet cheet to this issue to illustrate whats been done for those not keen on docs.

Here's how it goes.

#### Working directory
Regarding **working directory**, thats is now always the *current directory* from where `star.exe` is invoked, when starting from the command-line.

For
```
C:\foo\bar>star d:\apps\myapp.exe
```

the working directory is `c:\foo\bar`.

If you run from Visual Studio, based on an application created from `Starcounter application` project template, we'll set the **default** working directory to the **project directory**. Open `Project | Properties | Debug` tab and see the `Working directory` text box there. You are free to change it to your liking.

#### Resource directories
The important parts already documented [here](http://starcounter.io/guides/network/static-file-server/). Here's some hands-on:

In the code host, `Application` now expose a readonly `string[] ResourceDirectories` property. Read the docs for the semantics.

Here's a simple test program to see it in use.

```C#

using Starcounter;
using System;
using System.IO;

class Program {
    static void Main() {
        // Show the working directory and all resource directories
        // specified on the command line, in their qualified form
        Console.WriteLine("Working directory: {0}", Application.Current.WorkingDirectory);
        Console.WriteLine("Resource directories:");
        foreach (var resDir in Application.Current.ResourceDirectories) {
            Console.WriteLine(resDir);
        }

        // For each resource directory, we create a file within it. We
        // write the full path of the containing resource directory as
        // the content, and display in the console the name of the file.
        // With default setups, these files should thereafter be accessible
        // using localhost:8080/file[n]
        int file = 0;
        foreach (var resDir in Application.Current.ResourceDirectories) {
            var path = Path.Combine(resDir, "file" + file + ".html");
            File.WriteAllText(path, string.Format("I'm in {0}!", resDir));
            Console.WriteLine("Support /{0}", Path.GetFileName(path));
            file++;
        }
    }
}
```

It prints out the working directory and all resource directories *explicitly* given to the application when it start, once resolved. For each directory, it stores a file `file[n].html`, with the content that shows in what directory it reside. You can view these files by means of `localhost:8080/file[n]`.

As an example, putting the above into **app.cs** and running

```
C:\users\per\proj\issue3099>star --resourceDir="foo;foo\bar;c:\users\per" app.cs
```

would allow me to access `localhost:8080/file[n]` where [n] is 0-2.

In addition, `localhost:8080/staticcontentdir` still reveal *all* resource directories in use for a given port. Browsing that after starting the above application shows this:

```
http://localhost:8080/staticcontentdir
{"Items":[
 {"Port":8080,"Folder":"c:\\users\\per\\git\\starcounter\\develop\\level1\\bin\\debug\\clientfiles\\staticfiles"},
  {"Port":8080,"Folder":"c:\\users\\per\\proj\\issue3099\\foo"},
  {"Port":8080,"Folder":"c:\\users\\per\\proj\\issue3099\\foo\\bar"},
  {"Port":8080,"Folder":"c:\\users\\per"}
  ]
}
```

Now, if I create a `wwwroot` and rerun it, it reflect that.

```
C:\users\per\proj\issue3099>mkdir wwwroot
C:\users\per\proj\issue3099>star --resourceDir="foo;foo\bar;c:\users\per" app.cs

http://localhost:8080/staticcontentdir
{"Items":[
  {"Port":8080,"Folder":"c:\\users\\per\\git\\starcounter\\develop\\level1\\bin\\debug\\clientfiles\\staticfiles"},
  {"Port":8080,"Folder":"c:\\users\\per\\proj\\issue3099\\wwwroot"},
  {"Port":8080,"Folder":"c:\\users\\per\\proj\\issue3099\\foo"},
  {"Port":8080,"Folder":"c:\\users\\per\\proj\\issue3099\\foo\\bar"},
  {"Port":8080,"Folder":"c:\\users\\per"}
  ]
}
```

That reflecs the convention-based inclusion of ```wwwroot```. For this, yet another path is supported.

```
C:\users\per\proj\issue3099>mkdir src\app\wwwroot
C:\users\per\proj\issue3099>tree

Folder PATH listing
Volume serial number is D893-7418
C:.
├───src
│   └───app
│       └───wwwroot
└───wwwroot

C:\users\per\proj\issue3099>star --resourceDir="foo;foo\bar;c:\users\per" app.cs
http://localhost:8080/staticcontentdir
{"Items":[
  {"Port":8080,"Folder":"c:\\users\\per\\git\\starcounter\\develop\\level1\\bin\\debug\\clientfiles\\staticfiles"},
  {"Port":8080,"Folder":"c:\\users\\per\\proj\\issue3099\\wwwroot"},
  {"Port":8080,"Folder":"c:\\users\\per\\proj\\issue3099\\foo"},
  {"Port":8080,"Folder":"c:\\users\\per\\proj\\issue3099\\foo\\bar"},
  {"Port":8080,"Folder":"c:\\users\\per"}
  ]
}
```

Here, since both `wwwroot` and `src\[appname]\wwwroot` patterns make a hit, we just use the closest one, and emit a notice that the second one is ignored.

```
staradmin list --max=1 log

[14:54:38, Notice, Starcounter.Host (default)]
Application "app" contains both \wwwroot and \src\app\wwwroot. Ignoring the later.
```

Finally, illegal characters in resource directories are caught and we refuse to boot the app.

```
C:\users\per\proj\issue3099>star --resourceDir="foo>>>;foo\bar;c:\users\per" app.cs

ScErrBadArguments (SCERR1001): One or more arguments was invalid. Parameter --resourcedir contains an illegal element: 'foo>>>', error: Illegal characters in path.
Version: 2.0.0.
```

**Comment by ivarbergman** [comment]
Hi! I have a issue when using a relative resource directory.

    C:\Tunity\Project\Tunity-3\Tunity.Gantt>star --resourceDir="Client;..\Tunity.Common\Client" "bin\Debug\tunity_gantt.exe"

result in the following error message:

    Please check it existence and correct access rights. 
    Directory path: C:\Tunity\Project\Tunity-3\Tunity.Gantt\..\..\Tunity.Common\Client

It seems that it adds en extra `..\`

Here is the complete stack trace.

    Starcounter.DbException: ScErrFailingEntrypoint (SCERR4259): The user code entrypoint method raised an exception. ScErrCantAccessAppResourceDir (SCERR12020): Specified application resource directory is not accessible. Please check it existence and correct access rights. Directory path: C:\Tunity\Project\Tunity-3\Tunity.Gantt\..\..\Tunity.Common\Client
    Version: 2.1.656.
    Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR12020.
    Version: 2.1.656.
    Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR4259.
       at Starcounter.Hosting.Package.ExecuteEntryPoint(Application application) inC:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-106\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 583
       at Starcounter.Hosting.Package.ProcessWithinCurrentApplication(Application application) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-106\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 270
       at Starcounter.Hosting.Package.Process() in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-106\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 143
       at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-106\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 76

**Comment by careri** [comment]
Nice, makes much more sense now. Even I understand what's going on :tada: 

One problem though, providing the resourcedir from the debug tab in VS seems inconsistent:
Works:

    --resourceDir=..\..\wwwroot

Doesn't work (quotes seems to be thre problem):
```
--resourceDir="..\..\wwwroot"
--resourceDir="d:\dummy\testproject\wwwroot"
--resourceDir="..\\..\\wwwroot"
--resourceDir="d:\\dummy\\testprojec\t\wwwroot"
```

```
Error	9	One or more arguments was invalid. Parameter --resourcedir contains an illegal element: '"..\\..\\wwwroot"', error: Illegal characters in path. (ScErrBadArguments (SCERR1001))

(Double-click here for additional help)		0	0	
```

**Comment by per-samuelsson** [comment]
> Nice, makes much more sense now. Even I understand what's going on.

Haha, thanks. Glad you like it. :smiley:

> Doesn't work (quotes seems to be thre problem).

I see. I never tried with quotes actually. I'll see if I can verify it and open up an separate issue regarding that bug.

So you are already using it? Your deploy nightlys, do you? Cool.

**Comment by careri** [comment]
Yepp, nightlys. I was looking at your example for cmd:
```
C:\users\per\proj\issue3099>star --resourceDir="foo;foo\bar;c:\users\per" app.cs
```

**Comment by per-samuelsson** [comment]
> I was looking at your example for cmd:

You should always stick to the CLI. Never any problems there! :wink:

**Comment by careri** [comment]
Yepp, CLI is usally the fastest way of doing things. I'm all for hotkeys and minimizing mouse usage.
This is faster (one key):
F5 in VS

This is slower:
CTRL+SHIFT+B (build solution)
ALT+TAB to cmd (or Win+3 in my case as it's pinned to taskbar)
Up arrow to go to previous command
ALT+TAB to VS (or win+4 in my case)
CTRL+ALT+P (attach debugger)
type sccode
hit enter to attach

;)

**Comment by careri** [comment]
One more though. Shouldn't the default working folder for the project in VS be the output folder and not the project folder?

For most other project types hitting F5 (debug) means that the code in the output folder is executed

**Comment by andwah** [comment]
> Shouldn't the default working folder for the project in VS be the output folder and not the project folder?

 If we run our app and and modify some html then we do not want to rebuild the project , we just want to hit refresh in the browser.

**Comment by careri** [comment]
Thats what's the resourceDir argument is used for right? If I follow your conclusion about auto loading static resources and not doing any files in the project folder shouldn't be auto loaded either?

I suggest:
output=$(outputpath)
Arg=--resourceDir=$(MSBuildProjectDirectory)

This makes the app run in the output directory. What I the user want's to read som custom datafiles residing in the $(outputpath) how will he find them? Searching for a subfolder named bin\debug?

Rember that not everyone want's to use starcounter for html-applications. Maybe you should have two different project types?

**Comment by per-samuelsson** [comment]
> Any files in the project folder shouldn't be auto loaded either?

They aren't. If you read the docs, you'll see that we look for `wwwroot` as a *subfolder* to working directory. So the working directory (i.e project directory) is not exposed by default.

With that being said, the design with working dir = project directory is not mine, so I can't really say whats right or wrong. For me, the setup you suggest would be the most natural too. But there might be things relating to "app dev" I'm not aware of that need the current pattern.

> Maybe you should have two different project types?

Yeah, we've discussed that. For other reasons too. But we are not there yet. Focus is very much on "app dev", hence most stuff we do right now is for that.

> What I the user want's to read som custom datafiles residing in the $(outputpath) how will he find them? Searching for a subfolder named bin\debug?

No, he'll use `Path.GetDirectory(Application.Current.FilePath)` as I have instructed before. The working directory can not be trusted in any way we choose because 1) it can be specified to be whatever, and 2) if you run from the command-line, working directory is where you execute ´star` from.

So even though I agree with your suggestion, it doesn't really relate 1-1 to that last case.
